gulp-tasks/build-packages.js

```diff
@@ -1,204 +1,59 @@
-  return fs.remove(outputDirectory);
+    constants.PACKAGE_BUILD_DIRNAME);
+  return fse.remove(outputDirectory);
 };
 
 gulp.task('build-packages:clean', gulp.series(
-  packageRunnner('build-packages:clean', cleanPackages)
+    packageRunnner.wrapFunction(
+      'Cleaning Package',
+      packageRunnner.getPackages(),
+      cleanPackage
+    )
 ));
 
-// This will create one version of the tests for each buildType.
-// i.e. we'll have a browser build for no NODE_ENV and one for 'production'
-// NODE_ENV and the same for sw and node tests.
-const packageBuilds = constants.BUILD_TYPES.map((buildType) => {
-  return packageRunnner('build-package', buildPackage, buildType);
-});
+gulp.task('build-packages:build', (done) => {
+  const buildStreams = [];
 
-gulp.task('build-packages:build', gulp.series(packageBuilds));
+  const nodePackages = packageRunnner.getPackages('node');
+  if (nodePackages.length > 0) {
+    buildStreams.push(...packageRunnner.wrapFunction(
+      'Building Node Package',
+      nodePackages,
+      buildNodePackage
+    ));
+  }
 
-/*
- * This function will take an object and generate a friend export
- * object.
- * For example, { hello: world, example: { foo: foo } } will output:
- * "{
- *   hello: world,
- *   example: {
- *     foo,
- *   },
- * }"
- * @param {*} exportObject This is the object that needs to be converted
- * to a string.
- * @param {*} levels This value is just used for indenting the code to
- * ensure the final output is human readable and easy to understand.
- */
-const convertExportObject = (exportObject, levels = 0) => {
-  let outputStrings = [];
-  const keys = Object.keys(exportObject);
-  keys.forEach((key) => {
-    const value = exportObject[key];
-    if (typeof(value) === 'string') {
-      if (key === value) {
-        outputStrings.push(`${value}`);
-      } else {
-        outputStrings.push(`${key}: ${value}`);
-      }
-    } else {
-      const valueString = convertExportObject(value, levels + 1);
-      outputStrings.push(`${key}: ${valueString}`);
+  const browserPackages = packageRunnner.getPackages('browser');
+  if (browserPackages.length > 0) {
+    for (const buildType of constants.BUILD_TYPES) {
+      buildStreams.push(...packageRunnner.wrapFunction(
+        'Building Browser Package',
+        browserPackages,
+        buildBrowserPackage,
+        buildType
+      ));
     }
-  });
-  const padding = '  '.repeat(levels + 1);
-  const closePadding = '  '.repeat(levels);
-  const joinString = `,\n${padding}`;
-  return `{\n${padding}${outputStrings.join(joinString)}\n${closePadding}}`;
-};
+  }
 
-gulp.task('build-packages', gulp.series([
+  // Our various builds all return gulp streams. Kick them off and merge the
+  // result into a single stream that will emit 'end' when they have completed.
+  // Use that to signal the overall task completion.
+  return eventStream.merge(buildStreams.map((func) => func()))
```

Comment:
It seems like there *should* be a more straightforward way doing this in `gulp` 4, but after banging my head for a while, this was the only thing that I could come up with that would wait for the end of all the streams that are returned by the build functions.

Reply:
If you seperate this out into seperate tasks it'll be cleaner.

```js
gulp.task('build-packages:build', gulp.parallel('build-browser-packages', 'build-node-packages'));
```

Then create `gulp-tasks/build-browser-packages.js` and `gulp-tasks/build-node-packages.js`

Reply:
That seems to work!